File: crates/basilica-sdk-python/python/basilica/templates/model_size.py
"""
Model size estimation and GPU requirement calculation.

Estimates GPU memory requirements based on model name patterns.
Uses heuristics based on parameter count extracted from model names.
"""

import math
import re
from dataclasses import dataclass
from typing import Optional

# Default GPU memory for available GPUs (16GB for RTX A4000)
DEFAULT_GPU_MEMORY_GB = 16

# Regex pattern for extracting parameter count from model names
# Matches patterns like "7b", "70b", "0.5b", "1.5b", "7B", etc.
PARAM_PATTERN = re.compile(r"(\d+\.?\d*)b", re.IGNORECASE)


@dataclass
class GpuRequirements:
    """GPU requirements for a model."""

    gpu_count: int
    memory_gb: int
    recommended_gpu: str


def estimate_gpu_requirements(model: str) -> GpuRequirements:
    """
    Estimate GPU requirements based on model name.

    Uses heuristics based on parameter count extracted from model name.
    Falls back to model family detection if parameter count cannot be extracted.

    Args:
        model: HuggingFace model identifier (e.g., "meta-llama/Llama-2-7b")

    Returns:
        GpuRequirements with gpu_count, memory_gb, and recommended_gpu
    """
    model_lower = model.lower()

    # Extract parameter count from model name
    params = extract_param_count(model_lower)

    if params is not None:
        memory_gb = estimate_memory_from_params(params)
    else:
        memory_gb = estimate_from_model_family(model_lower)

    gpu_count = calculate_gpu_count(memory_gb, DEFAULT_GPU_MEMORY_GB)
    recommended_gpu = recommend_gpu(memory_gb)

    return GpuRequirements(
        gpu_count=gpu_count,
        memory_gb=memory_gb,
        recommended_gpu=recommended_gpu,
    )


def extract_param_count(model: str) -> Optional[float]:
    """
    Extract billion parameter count from model name.

    Patterns: "7b", "70b", "0.5b", "1.5b", "7B", etc.

    Args:
        model: Lowercase model name

    Returns:
        Parameter count in billions, or None if not found
    """
    match = PARAM_PATTERN.search(model)
    if match:
        return float(match.group(1))
    return None


def estimate_memory_from_params(params_billions: float) -> int:
    """
    Estimate VRAM needed in GB from parameter count.

    Rule of thumb: ~2GB per billion parameters for FP16.
    Add 20% overhead for KV cache and runtime.

    Args:
        params_billions: Model size in billions of parameters

    Returns:
        Estimated VRAM in GB, rounded to nearest 8GB
    """
    # Base memory: 2GB per billion params * 1.2 overhead
    base = params_billions * 2.0 * 1.2
    # Round up to nearest 8GB
    return ((int(base) + 7) // 8) * 8


def estimate_from_model_family(model: str) -> int:
    """
    Estimate memory from model family when parameter count is not available.

    Args:
        model: Lowercase model name

    Returns:
        Estimated VRAM in GB
    """
    # Check for known large models
    if any(
        x in model
        for x in ["llama-2-70b", "llama-70b", "mixtral", "qwen-72b"]
    ):
        return 160  # ~70B params

    if any(
        x in model
        for x in ["llama-2-13b", "llama-13b", "codellama-34b"]
    ):
        return 32  # ~13-34B params

    if any(
        x in model
        for x in ["llama-2-7b", "llama-7b", "mistral-7b", "qwen-7b"]
    ):
        return 16  # ~7B params

    if any(
        x in model
        for x in ["phi-2", "gemma-2b", "tinyllama", "qwen3-0.6b", "qwen2.5-0.5b"]
    ):
        return 8  # Small models

    # Default: assume medium-sized model
    return 16


def calculate_gpu_count(required_memory_gb: int, gpu_memory_gb: int = 16) -> int:
    """
    Calculate number of GPUs needed.

    Args:
        required_memory_gb: Required VRAM in GB
        gpu_memory_gb: Available memory per GPU (default: 16GB)

    Returns:
        Number of GPUs needed (clamped to 1-8)
    """
    count = math.ceil(required_memory_gb / gpu_memory_gb)
    return min(max(count, 1), 8)  # Clamp 1-8


def recommend_gpu(memory_gb: int) -> str:
    """
    Recommend GPU based on memory requirements.

    Args:
        memory_gb: Required VRAM in GB

    Returns:
        Recommended GPU model name
    """
    if memory_gb <= 16:
        return "NVIDIA-RTX-A4000"
    elif memory_gb <= 40:
        return "A100-40GB"
    elif memory_gb <= 80:
        return "A100-80GB"
    else:
        return "H100"

File: crates/basilica-cli/src/cli/handlers/deploy/templates/model_size.rs
//! Model size estimation and GPU requirement calculation
//!
//! Provides sensible defaults based on model name patterns.
//! Users can override with explicit CLI flags (--gpu, --memory, --gpu-model).

use regex::Regex;
use std::sync::LazyLock;

/// GPU requirements for a model
#[derive(Debug, Clone)]
pub struct GpuRequirements {
    /// Number of GPUs required
    pub gpu_count: u32,
    /// Estimated GPU memory required in GB
    pub memory_gb: u32,
    /// Recommended GPU model
    pub recommended_gpu: String,
}

/// Default GPU memory per device (80GB for A100/H100)
const DEFAULT_GPU_MEMORY_GB: u32 = 80;

/// Default memory estimate when parameter count cannot be extracted
const DEFAULT_MEMORY_GB: u32 = 16;

/// Regex pattern for extracting parameter count from model names
/// Matches patterns like "7b", "70b", "0.5b", "1.5b", "7B", "235B", etc.
static PARAM_REGEX: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(\d+\.?\d*)[bB]").expect("Invalid regex pattern"));

/// Estimate GPU requirements based on model name
///
/// Extracts parameter count from model name and estimates memory needs.
/// Returns sensible defaults that users can override with CLI flags.
pub fn estimate_gpu_requirements(model: &str) -> GpuRequirements {
    let memory_gb = extract_param_count(model)
        .map(estimate_memory_from_params)
        .unwrap_or(DEFAULT_MEMORY_GB);

    let gpu_count = calculate_gpu_count(memory_gb, DEFAULT_GPU_MEMORY_GB);
    let recommended_gpu = recommend_gpu(memory_gb);

    GpuRequirements {
        gpu_count,
        memory_gb,
        recommended_gpu,
    }
}

/// Extract billion parameter count from model name
fn extract_param_count(model: &str) -> Option<f32> {
    PARAM_REGEX
        .captures(model)
        .and_then(|cap| cap.get(1))
        .and_then(|m| m.as_str().parse::<f32>().ok())
}

/// Estimate VRAM needed in GB from parameter count
///
/// Rule of thumb: ~2GB per billion parameters for FP16
/// Add 20% overhead for KV cache and runtime
fn estimate_memory_from_params(params_billions: f32) -> u32 {
    let base = params_billions * 2.0 * 1.2;
    // Round up to nearest 8GB
    (base as u32).div_ceil(8) * 8
}

/// Calculate number of GPUs needed
fn calculate_gpu_count(required_memory_gb: u32, gpu_memory_gb: u32) -> u32 {
    let count = (required_memory_gb as f32 / gpu_memory_gb as f32).ceil() as u32;
    count.clamp(1, 8)
}

/// Recommend GPU based on memory requirements
fn recommend_gpu(memory_gb: u32) -> String {
    if memory_gb <= 80 {
        "A100".to_string()
    } else {
        "H100".to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_param_count() {
        assert!((extract_param_count("llama-2-7b").unwrap() - 7.0).abs() < 0.001);
        assert!((extract_param_count("Qwen3-0.6B").unwrap() - 0.6).abs() < 0.001);
        assert!((extract_param_count("Llama-2-70b").unwrap() - 70.0).abs() < 0.001);
        assert!((extract_param_count("Qwen3-235B").unwrap() - 235.0).abs() < 0.001);
        assert!(extract_param_count("phi-2").is_none()); // No 'b' suffix
    }

    #[test]
    fn test_estimate_memory_from_params() {
        // 7B: 7 * 2 * 1.2 = 16.8 -> 16GB (truncated then rounded up to 8)
        assert_eq!(estimate_memory_from_params(7.0), 16);
        // 0.5B: 0.5 * 2 * 1.2 = 1.2 -> 8GB
        assert_eq!(estimate_memory_from_params(0.5), 8);
        // 70B: 70 * 2 * 1.2 = 168 -> 168GB
        assert_eq!(estimate_memory_from_params(70.0), 168);
        // 235B: 235 * 2 * 1.2 = 564 -> 568GB
        assert_eq!(estimate_memory_from_params(235.0), 568);
    }

    #[test]
    fn test_calculate_gpu_count() {
        assert_eq!(calculate_gpu_count(16, 80), 1);
        assert_eq!(calculate_gpu_count(80, 80), 1);
        assert_eq!(calculate_gpu_count(160, 80), 2);
        assert_eq!(calculate_gpu_count(320, 80), 4);
        assert_eq!(calculate_gpu_count(640, 80), 8);
        assert_eq!(calculate_gpu_count(1000, 80), 8); // Capped at 8
    }

    #[test]
    fn test_estimate_gpu_requirements() {
        // Small model
        let reqs = estimate_gpu_requirements("Qwen/Qwen2.5-0.5B-Instruct");
        assert_eq!(reqs.gpu_count, 1);
        assert_eq!(reqs.memory_gb, 8);
        assert_eq!(reqs.recommended_gpu, "A100");

        // 7B model
        let reqs = estimate_gpu_requirements("meta-llama/Llama-2-7b");
        assert_eq!(reqs.gpu_count, 1);
        assert_eq!(reqs.recommended_gpu, "A100");

        // 70B model
        let reqs = estimate_gpu_requirements("meta-llama/Llama-2-70b");
        assert!(reqs.gpu_count >= 2);
        assert_eq!(reqs.recommended_gpu, "H100");

        // 235B model (extracts from name)
        let reqs = estimate_gpu_requirements("Qwen/Qwen3-235B-A22B");
        assert!(reqs.gpu_count >= 4);
        assert_eq!(reqs.recommended_gpu, "H100");
    }

    #[test]
    fn test_unknown_model_uses_default() {
        let reqs = estimate_gpu_requirements("some-unknown-model");
        assert_eq!(reqs.memory_gb, DEFAULT_MEMORY_GB);
        assert_eq!(reqs.gpu_count, 1);
    }
}
